### PR TITLE
Add batch query, rate limiter and improve logging

### LINKED ISSUE #15: Use a batch query to scan up to 1K dependencies at once
The OSV API has an endpoint to submit batch queries and get known vulnerabilities for up to 1K dependencies at once. Leveraging this feature would let us create a list of vulnerabilities in a single request, and get details for each of them only when needed (e.g. when scrolling the list in a grid).

See: https://google.github.io/osv.dev/post-v1-querybatch/

### PR DESCRIPTION
**Pull request by tamasmak**
- Adds batch query to get known vulnerabilities for up to 1K dependencies at once
- Adds rate limiter to limit the calls per second to the OSV API. The rate per second is configurable.
- Improves exception handling and logging

Closes #15

### CODE REVIEW

File: appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/AppSecScanEvent.java [hunk 1]
```diff
@@ -20,6 +23,9 @@
  */
 public class AppSecScanEvent extends EventObject {
 
+    private static final Logger LOGGER = LoggerFactory
+            .getLogger(AppSecScanEvent.class);
+
```
Comment (heruan) [inline appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/AppSecScanEvent.java:28]: I would avoid having loggers in event objects, as they're purpose is to contain data not doing anything.

**Review by heruan — CHANGES_REQUESTED** [review]
I added some comments to address, but overall looks good!

Now we need tests for the new logic added for rate limiting, but I don't think we have any test for the OSV service and client at all so it's a good time to add those before merging this.

**Review by heruan — COMMENTED** [review]
I just pushed a commit addressing my own comments, in particular simplifying the use of the stream API in [`OpenSourceVulnerabilityService`](https://github.com/vaadin/appsec-kit/pull/40/commits/297900187dc946a7e9bf92178723179128925fe6#diff-e73925028a10e5e99783c84d7ac47ecbc7a3a615079d99799f22ee3aaea10341)

Other changes include:
- move the rate limiter to the client
- return non-optional query results
- remove the logger from the scan event

File: appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityClient.java [hunk 2]
```diff
@@ -41,14 +46,18 @@ Optional<QueryBatchResponse> queryBatch(QueryBatchRequestPayload payload) {
...
+                    "Failed to write OSV API query batch request payload", e);
         }
 
         try (InputStream inputStream = conn.getInputStream()) {
-            return Optional.of(
-                    mapper.readValue(inputStream, QueryBatchResponse.class));
+            QueryBatchResponse response = mapper.readValue(inputStream,
+                    QueryBatchResponse.class);
+            LOGGER.debug("OSV API batch query for " + payload);
+            return Optional.of(response);
```
[the remark this answers is not in the document]
Comment (tamasmak) [reply appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityClient.java]: Yes, that's true. Probably left there from some early phase.

File: appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityClient.java [hunk 3]
```diff
@@ -58,10 +67,13 @@ Optional<OpenSourceVulnerability> queryVulnerability(
         HttpURLConnection conn = createHttpURLConnection(urlWithId, "GET");
 
         try (InputStream inputStream = conn.getInputStream()) {
-            return Optional.of(mapper.readValue(inputStream,
-                    OpenSourceVulnerability.class));
+            OpenSourceVulnerability vulnerability = mapper
+                    .readValue(inputStream, OpenSourceVulnerability.class);
+            LOGGER.debug("OSV API vulnerability query for " + vulnerabilityId);
+            return Optional.of(vulnerability);
```
[the remark this answers is not in the document]
Comment (tamasmak) [reply appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityClient.java]: Same comment here.

File: appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityService.java [hunk 4]
```diff
@@ -32,7 +34,18 @@
...
+            .getLogger(OpenSourceVulnerabilityService.class);
+
+    private static final int COMPONENT_BATCH_SIZE = 1000;
+
+    private final OpenSourceVulnerabilityClient osvClient;
+    private final RateLimiter rateLimiter;
+
+    OpenSourceVulnerabilityService(int ratePerSecond) {
+        osvClient = new OpenSourceVulnerabilityClient();
+        rateLimiter = new RateLimiter(ratePerSecond);
```
[the remark this answers is not in the document]
Comment (tamasmak) [reply appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityService.java]: I agree.

File: appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityService.java [hunk 5]
```diff
@@ -47,33 +60,61 @@ List<OpenSourceVulnerability> getVulnerabilities(
...
+
+            batchResponse.ifPresent(response -> vulnerabilityIds.addAll(Arrays
+                    .stream(response.getResults())
+                    .filter(vulnerabilityIdArray -> vulnerabilityIdArray
+                            .getVulnerabilityIds() != null)
+                    .flatMap(vulnerabilityIdArray -> Arrays
+                            .stream(vulnerabilityIdArray.getVulnerabilityIds())
+                            .filter(Objects::nonNull))
+                    .map(VulnerabilityId::getId).collect(Collectors.toList())));
+        });
```
[the remark this answers is not in the document]
Comment (tamasmak) [reply appsec-kit-backend/src/main/java/com/vaadin/appsec/backend/OpenSourceVulnerabilityService.java]: Thanks. The solution you committed is really simple and nice.

**Review by heruan — CHANGES_REQUESTED** [review]
I do not clearly get what it's being tested in `OpenSourceVulnerabilityClientTest`: I would expect tests to validate queries such as verify they're using the correct method, headers and content.

This way of mocking the whole URL API is clever, but I'd avoid that: we want to test our own API, and using URL is an implementation detail. Better go with something like `MockServer` to test that the client sends out correct HTTP requests regardless of what we're using internally to send those.

**Review by heruan — APPROVED** [review]
I pushed a commit that uses `MockServer` to verify the OSV client. Tests now verify the correct use of the API by checking the request method, and content-type. For the batch query request, we should also verify that the request body is valid, but let's do this later.

**Pull request merged**